File: ios/runanywhere-sdks-latest/sdk/runanywhere-web/packages/core/src/Foundation/SDKLogger.ts
/**
 * RunAnywhere Web SDK - Logger
 *
 * Logging system matching the pattern across all SDKs.
 * Routes to console.* methods in the browser.
 */

// `LogLevel` and `LoggingConfiguration` are the generated proto types
// (idl/logging.proto). The proto `LogLevel` mirrors the C ABI rac_log_level_t
// (LOG_LEVEL_TRACE=0 .. LOG_LEVEL_FATAL=5), which is numerically identical to
// the Web SDK's previous hand-written enum, so adopting it is value-stable.
// Severity ordering ("larger = more severe") holds, so the `<` comparisons in
// `log()` keep working. The hand-written `LoggingConfiguration` was a subset of
// the generated one (which additionally carries enableRemoteLogging /
// includeSourceLocation / includeDeviceMetadata); the console logger only acts
// on the three fields it supported before and ignores the rest.
import {
  LogLevel,
  LoggingConfiguration as LoggingConfigurationProto,
  type LoggingConfiguration,
} from '@runanywhere/proto-ts/logging';
import { SDKEnvironment } from '@runanywhere/proto-ts/model_types';

export { LogLevel };
export type { LoggingConfiguration };

/**
 * Environment presets — Swift parity: `RALoggingConfiguration.development /
 * .staging / .production` (SDKLogger.swift:41-66). Dev logs at debug with
 * local logging on; production logs warnings only with local logging off.
 */
export function loggingConfigurationForEnvironment(
  environment: SDKEnvironment,
): LoggingConfiguration {
  switch (environment) {
    case SDKEnvironment.SDK_ENVIRONMENT_STAGING:
      return LoggingConfigurationProto.fromPartial({
        enableLocalLogging: true,
        minLogLevel: LogLevel.LOG_LEVEL_INFO,
        includeSourceLocation: false,
        includeDeviceMetadata: true,
        enableRemoteLogging: false,
      });
    case SDKEnvironment.SDK_ENVIRONMENT_PRODUCTION:
      return LoggingConfigurationProto.fromPartial({
        enableLocalLogging: false,
        minLogLevel: LogLevel.LOG_LEVEL_WARNING,
        includeSourceLocation: false,
        includeDeviceMetadata: true,
        enableRemoteLogging: false,
      });
    default:
      return LoggingConfigurationProto.fromPartial({
        enableLocalLogging: true,
        minLogLevel: LogLevel.LOG_LEVEL_DEBUG,
        includeSourceLocation: false,
        includeDeviceMetadata: false,
        enableRemoteLogging: false,
      });
  }
}

/** Map LogLevel to RACommons rac_log_level_t values. Values are already
 * C-ABI-aligned, so this is the identity mapping. */
export const LOG_LEVEL_TO_RAC: Record<LogLevel, number> = {
  [LogLevel.LOG_LEVEL_TRACE]: 0,
  [LogLevel.LOG_LEVEL_DEBUG]: 1,
  [LogLevel.LOG_LEVEL_INFO]: 2,
  [LogLevel.LOG_LEVEL_WARNING]: 3,
  [LogLevel.LOG_LEVEL_ERROR]: 4,
  [LogLevel.LOG_LEVEL_FATAL]: 5,
  [LogLevel.UNRECOGNIZED]: -1,
};

/**
 * Represents a destination that can receive log entries.
 * Mirrors Swift `LogDestination` protocol (SDKLogger.swift:28-32).
 */
export interface LogDestination {
  readonly identifier: string;
  /** Whether the destination can currently accept entries (e.g. a remote
   * sink that is not configured yet). Treated as `true` when omitted —
   * mirrors Swift's `isAvailable` gate in the destination fan-out. */
  readonly isAvailable?: boolean;
  write(level: LogLevel, category: string, message: string): void;
  flush(): void;
}

export class SDKLogger {
  private static _level: LogLevel = LogLevel.LOG_LEVEL_INFO;
  private static _enabled = true;
  private static _extraDestinations: Map<string, LogDestination> = new Map();

  private readonly category: string;

  constructor(category: string) {
    this.category = category;
  }

  // -------------------------------------------------------------------------
  // Static configuration surface — mirrors Swift Logging.shared.*
  // -------------------------------------------------------------------------

  static get level(): LogLevel {
    return SDKLogger._level;
  }

  static set level(level: LogLevel) {
    SDKLogger._level = level;
  }

  static get enabled(): boolean {
    return SDKLogger._enabled;
  }

  static set enabled(value: boolean) {
    SDKLogger._enabled = value;
  }

  /** Apply a full logging configuration in one call. Mirrors Swift `Logging.shared.configure(_:)`. */
  static configure(config: LoggingConfiguration): void {
    SDKLogger._enabled = config.enableLocalLogging;
    SDKLogger._level = config.minLogLevel;
  }

  /**
   * Apply configuration based on SDK environment. Mirrors Swift
   * `Logging.shared.applyEnvironmentConfiguration(_:)` (SDKLogger.swift:300).
   */
  static applyEnvironmentConfiguration(environment: SDKEnvironment): void {
    SDKLogger.configure(loggingConfigurationForEnvironment(environment));
  }

  /** Enable or disable local console output. Mirrors Swift `Logging.shared.setLocalLoggingEnabled(_:)`. */
  static setLocalLoggingEnabled(enabled: boolean): void {
    SDKLogger._enabled = enabled;
  }

  /** Set minimum log level. Mirrors Swift `Logging.shared.setMinLogLevel(_:)`. */
  static setMinLogLevel(level: LogLevel): void {
    SDKLogger._level = level;
  }

  /** Register a custom log destination. Mirrors Swift `Logging.shared.addDestination(_:)`. */
  static addDestination(destination: LogDestination): void {
    SDKLogger._extraDestinations.set(destination.identifier, destination);
  }

  /** Flush all registered destinations. Mirrors Swift `Logging.shared.flush()`. */
  static flush(): void {
    for (const destination of SDKLogger._extraDestinations.values()) {
      try {
        destination.flush();
      } catch { /* Swallow flush errors so logging never crashes the SDK. */ }
    }
  }

  debug(message: string): void {
    this.log(LogLevel.LOG_LEVEL_DEBUG, message);
  }

  info(message: string): void {
    this.log(LogLevel.LOG_LEVEL_INFO, message);
  }

  warning(message: string): void {
    this.log(LogLevel.LOG_LEVEL_WARNING, message);
  }

  error(message: string): void {
    this.log(LogLevel.LOG_LEVEL_ERROR, message);
  }

  /**
   * Log at FATAL severity. Mirrors Swift `SDKLogger.fault(_:metadata:)`
   * (SDKLogger.swift:337-339), which routes to `RALogLevel.fatal`. Metadata is
   * folded into the line Swift-console-style (` | key=value, key=value`).
   */
  fault(message: string, metadata?: Record<string, unknown>): void {
    const entries = metadata ? Object.entries(metadata) : [];
    const suffix = entries.length > 0
      ? ` | ${entries.map(([key, value]) => `${key}=${String(value)}`).join(', ')}`
      : '';
    this.log(LogLevel.LOG_LEVEL_FATAL, `${message}${suffix}`);
  }

  private log(level: LogLevel, message: string): void {
    const shouldLog = SDKLogger._enabled || SDKLogger._extraDestinations.size > 0;
    if (level < SDKLogger._level || !shouldLog) {
      return;
    }

    if (SDKLogger._enabled) {
      const prefix = `[RunAnywhere:${this.category}]`;

      switch (level) {
        case LogLevel.LOG_LEVEL_TRACE:
        case LogLevel.LOG_LEVEL_DEBUG:
          console.debug(prefix, message);
          break;
        case LogLevel.LOG_LEVEL_INFO:
          console.info(prefix, message);
          break;
        case LogLevel.LOG_LEVEL_WARNING:
          console.warn(prefix, message);
          break;
        case LogLevel.LOG_LEVEL_ERROR:
        case LogLevel.LOG_LEVEL_FATAL:
          console.error(prefix, message);
          break;
        case LogLevel.UNRECOGNIZED:
          console.error(prefix, message);
          break;
      }
    }

    for (const destination of SDKLogger._extraDestinations.values()) {
      // Swift parity (SDKLogger.swift:183): skip unavailable destinations;
      // an omitted isAvailable defaults to true.
      if (destination.isAvailable === false) continue;
      try {
        destination.write(level, this.category, message);
      } catch { /* Swallow destination errors so logging never crashes the SDK. */ }
    }
  }
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-kotlin/src/main/kotlin/com/runanywhere/sdk/infrastructure/logging/SDKLogger.kt
package com.runanywhere.sdk.infrastructure.logging

import ai.runanywhere.proto.v1.LogEntry
import ai.runanywhere.proto.v1.LogLevel
import ai.runanywhere.proto.v1.LoggingConfiguration
import com.runanywhere.sdk.kotlin.BuildConfig
import com.runanywhere.sdk.public.configuration.SDKEnvironment
import com.runanywhere.sdk.utils.getCurrentTimeMillis
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

// LOG LEVEL / LOG ENTRY / LOGGING CONFIGURATION (proto-canonical)
//
// LogLevel, LogEntry and LoggingConfiguration are now the proto-generated
// types in ai.runanywhere.proto.v1. LogLevel values: LOG_LEVEL_TRACE(0),
// LOG_LEVEL_DEBUG(1), LOG_LEVEL_INFO(2), LOG_LEVEL_WARNING(3),
// LOG_LEVEL_ERROR(4), LOG_LEVEL_FATAL(5). Ordered by severity, larger value =
// more severe. Filtering: log iff `level.value >= minLogLevel.value`.
//
// LogEntry's extra Kotlin context (file/line/function/error_code/model_id/
// framework) maps onto the proto fields `file_`, `line`, `function`,
// `error_code`, `model_id`, `framework`; `timestamp_unix_ms` carries the
// epoch-millis timestamp. Proto fields use scalar defaults ("" / 0) instead of
// nullable — empty/zero means "unset".

// Per-environment LoggingConfiguration presets. The proto LoggingConfiguration
// is a flat message with no companion helpers, so the development/staging/
// production factories live here.

internal object LoggingConfigurationPresets {
    /** Development: console + debug level, no device metadata. */
    val development =
        LoggingConfiguration(
            enable_local_logging = true,
            min_log_level = LogLevel.LOG_LEVEL_DEBUG,
            include_device_metadata = false,
        )

    /** Staging: console + info level, device metadata on. */
    val staging =
        LoggingConfiguration(
            enable_local_logging = true,
            min_log_level = LogLevel.LOG_LEVEL_INFO,
            include_device_metadata = true,
        )

    /** Production: warning+ only, console off, device metadata on. */
    val production =
        LoggingConfiguration(
            enable_local_logging = false,
            min_log_level = LogLevel.LOG_LEVEL_WARNING,
            include_device_metadata = true,
        )

    fun forEnvironment(environment: SDKEnvironment): LoggingConfiguration =
        when (environment) {
            SDKEnvironment.SDK_ENVIRONMENT_DEVELOPMENT -> development
            SDKEnvironment.SDK_ENVIRONMENT_STAGING -> staging
            SDKEnvironment.SDK_ENVIRONMENT_PRODUCTION -> production
            SDKEnvironment.SDK_ENVIRONMENT_UNSPECIFIED -> development
        }
}

// Log destination protocol

/**
 * Protocol for log output destinations (Console, remote services, etc.).
 * Matches Swift SDK LogDestination protocol.
 */
interface LogDestination {
    /** Unique identifier for this destination */
    val identifier: String

    /** Whether this destination is available for writing */
    val isAvailable: Boolean

    /** Write a log entry to this destination */
    fun write(entry: LogEntry)

    /** Flush any buffered entries */
    fun flush()
}

// Logging (central service)

/**
 * Central logging service that routes logs to multiple destinations.
 * Thread-safe using Mutex for state management.
 * Matches Swift SDK Logging class.
 */
object Logging {
    private val mutex = Mutex()

    // Thread-safe state
    private var _configuration: LoggingConfiguration = LoggingConfigurationPresets.development
    private val _destinations: MutableList<LogDestination> = mutableListOf()

    // Bridge callback for forwarding logs to runanywhere-commons
    private var commonsLogBridge: ((LogEntry) -> Unit)? = null

    /**
     * Current logging configuration.
     */
    var configuration: LoggingConfiguration
        get() = _configuration
        set(value) {
            _configuration = value
        }

    /**
     * List of registered log destinations.
     */
    val destinations: List<LogDestination>
        get() = _destinations.toList()

    // Configuration

    /**
     * Configure the logging system.
     */
    fun configure(config: LoggingConfiguration) {
        _configuration = config
    }

    /**
     * Apply configuration based on SDK environment.
     */
    fun applyEnvironmentConfiguration(environment: SDKEnvironment) {
        configure(LoggingConfigurationPresets.forEnvironment(environment))
    }

    /**
     * Set whether local logging is enabled.
     */
    fun setLocalLoggingEnabled(enabled: Boolean) {
        _configuration = _configuration.copy(enable_local_logging = enabled)
    }

    /**
     * Set the minimum log level.
     */
    fun setMinLogLevel(level: LogLevel) {
        _configuration = _configuration.copy(min_log_level = level)
    }

    /**
     * Set whether to include source location in logs.
     */
    fun setIncludeSourceLocation(include: Boolean) {
        _configuration = _configuration.copy(include_source_location = include)
    }

    /**
     * Set whether to include device metadata in logs.
     */
    fun setIncludeDeviceMetadata(include: Boolean) {
        _configuration = _configuration.copy(include_device_metadata = include)
    }

    /**
     * Hook for delegating metadata redaction policy to the canonical C++
     * implementation `rac_log_metadata_should_redact`. Set by the platform
     * layer (jvmAndroidMain) once the native library is loaded, so Kotlin
     * and C++ logs share the same sensitive-substring list. When unset
     * (e.g. native library not yet loaded), `sanitizeMetadata` falls back
     * to the hardcoded `sensitivePatterns` list below.
     */
    var shouldRedactPolicy: ((String) -> Boolean)? = null
        internal set

    /**
     * Set the bridge callback for forwarding logs to runanywhere-commons.
     * This enables integration with the C/C++ logging system.
     */
    fun setCommonsLogBridge(bridge: ((LogEntry) -> Unit)?) {
        commonsLogBridge = bridge
    }

    // Core logging

    /**
     * Log a message with optional metadata.
     */
    fun log(
        level: LogLevel,
        category: String,
        message: String,
        metadata: Map<String, Any?>? = null,
        file: String? = null,
        line: Int? = null,
        function: String? = null,
        errorCode: Int? = null,
        modelId: String? = null,
        framework: String? = null,
    ) {
        val config = _configuration

        // Check if level meets minimum threshold. proto LogLevel severity:
        // LOG_LEVEL_TRACE(0) = least severe, LOG_LEVEL_FATAL(5) = most severe.
        // A log is emitted iff its severity is at least the configured
        // threshold (mirrors Swift `level >= config.minLogLevel`).
        if (level.value < config.min_log_level.value) return

        // Check if any logging is enabled
        if (!config.enable_local_logging && !config.enable_remote_logging && _destinations.isEmpty()) return

        // Create log entry. Proto LogEntry uses scalar defaults ("" / 0) for
        // unset fields rather than nullables.
        val includeSource = config.include_source_location
        val entryMetadata = enrichedMetadata(metadata, config.include_device_metadata)
        val entry =
            LogEntry(
                timestamp_unix_ms = getCurrentTimeMillis(),
                level = level,
                category = category,
                message = message,
                metadata = entryMetadata,
                file_ = if (includeSource) file.orEmpty() else "",
                line = if (includeSource) (line ?: 0) else 0,
                function = if (includeSource) function.orEmpty() else "",
                error_code = errorCode ?: 0,
                model_id = modelId.orEmpty(),
                framework = framework.orEmpty(),
            )

        // Write to console if local logging enabled
        if (config.enable_local_logging) {
            printToConsole(entry)
        }

        // Forward to runanywhere-commons bridge if set
        commonsLogBridge?.invoke(entry)

        // Write to all registered destinations (snapshot to avoid concurrent modification)
        for (destination in _destinations.toList()) {
            if (destination.isAvailable) {
                destination.write(entry)
            }
        }
    }

    // Destination management

    /**
     * Add a log destination.
     */
    suspend fun addDestination(destination: LogDestination) {
        mutex.withLock {
            if (_destinations.none { it.identifier == destination.identifier }) {
                _destinations.add(destination)
            }
        }
    }

    /**
     * Add a log destination (non-suspending version).
     */
    @Synchronized
    fun addDestinationSync(destination: LogDestination) {
        if (_destinations.none { it.identifier == destination.identifier }) {
            _destinations.add(destination)
        }
    }

    /**
     * Remove a log destination.
     */
    suspend fun removeDestination(destination: LogDestination) {
        mutex.withLock {
            _destinations.removeAll { it.identifier == destination.identifier }
        }
    }

    /**
     * Remove a log destination (non-suspending version).
     */
    fun removeDestinationSync(destination: LogDestination) {
        _destinations.removeAll { it.identifier == destination.identifier }
    }

    /**
     * Flush all destinations.
     */
    fun flush() {
        for (destination in _destinations) {
            destination.flush()
        }
    }

    // Private helpers

    private fun printToConsole(entry: LogEntry) {
        val levelIndicator =
            when (entry.level) {
                LogLevel.LOG_LEVEL_TRACE -> "[TRACE]"
                LogLevel.LOG_LEVEL_DEBUG -> "[DEBUG]"
                LogLevel.LOG_LEVEL_INFO -> "[INFO]"
                LogLevel.LOG_LEVEL_WARNING -> "[WARN]"
                LogLevel.LOG_LEVEL_ERROR -> "[ERROR]"
                LogLevel.LOG_LEVEL_FATAL -> "[FATAL]"
            }

        val output =
            buildString {
                append(levelIndicator)
                append(" [")
                append(entry.category)
                append("] ")
                append(entry.message)

                // Add metadata if present
                entry.metadata.takeIf { it.isNotEmpty() }?.let { meta ->
                    append(" | ")
                    append(meta.entries.joinToString(", ") { "${it.key}=${it.value}" })
                }

                // Add source location if present (proto: "" / 0 means unset)
                if (entry.file_.isNotEmpty() || entry.function.isNotEmpty()) {
                    append(" @ ")
                    if (entry.file_.isNotEmpty()) append(entry.file_)
                    if (entry.line != 0) append(":${entry.line}")
                    if (entry.function.isNotEmpty()) append(" in ${entry.function}")
                }

                // Add error code if present
                if (entry.error_code != 0) append(" [code=${entry.error_code}]")

                // Add model info if present
                if (entry.model_id.isNotEmpty() || entry.framework.isNotEmpty()) {
                    append(" [")
                    if (entry.model_id.isNotEmpty()) append("model=${entry.model_id}")
                    if (entry.model_id.isNotEmpty() && entry.framework.isNotEmpty()) append(", ")
                    if (entry.framework.isNotEmpty()) append("framework=${entry.framework}")
                    append("]")
                }
            }

        println(output)
    }

    // Metadata sanitization

    // Fallback sensitive-substring list used when the canonical C++ policy
    // (`rac_log_metadata_should_redact`) is not yet reachable — e.g. before
    // the native library is loaded or in non-JNI test contexts. Keep in
    // lockstep with `sdk/runanywhere-commons/src/infrastructure/logging/
    // log_redact.cpp` (kSensitiveSubstrings) and Swift `SDKLogger.swift`.
    private val sensitivePatterns = listOf("key", "secret", "password", "token", "auth", "credential")

    private const val DEVICE_MODEL_KEY = "device_model"
    private const val OS_VERSION_KEY = "os_version"
    private const val PLATFORM_KEY = "platform"

    /**
     * Determines whether a metadata key should be redacted. Delegates to the
     * canonical C++ policy via [shouldRedactPolicy] when set; otherwise falls
     * back to the hardcoded [sensitivePatterns] list. Any error thrown by the
     * policy hook (e.g. `UnsatisfiedLinkError` if the native library is
     * unavailable) is swallowed and the fallback list is used so logging never
     * fails on metadata sanitization.
     */
    private fun shouldRedact(key: String): Boolean {
        val policy = shouldRedactPolicy
        if (policy != null) {
            try {
                return policy(key)
            } catch (_: Throwable) {
                // Fall through to hardcoded list — keeps logging resilient if
                // the native bridge is unavailable.
            }
        }
        val lowercasedKey = key.lowercase()
        return sensitivePatterns.any { lowercasedKey.contains(it) }
    }

    @Suppress("UNCHECKED_CAST")
    private fun sanitizeMetadata(metadata: Map<String, Any?>?): Map<String, String>? {
        if (metadata == null) return null

        return metadata.mapValues { (key, value) ->
            when {
                shouldRedact(key) -> "[REDACTED]"
                value is Map<*, *> -> sanitizeMetadata(value as? Map<String, Any?>)?.toString() ?: "{}"
                else -> value?.toString() ?: "null"
            }
        }
    }

    private fun enrichedMetadata(
        metadata: Map<String, Any?>?,
        includeDeviceMetadata: Boolean,
    ): Map<String, String> {
        val sanitized = sanitizeMetadata(metadata).orEmpty()
        if (!includeDeviceMetadata) return sanitized
        return sanitized + deviceMetadata()
    }

    private fun deviceMetadata(): Map<String, String> =
        mapOf(
            DEVICE_MODEL_KEY to reflectedAndroidString("android.os.Build", "MODEL", "unknown"),
            OS_VERSION_KEY to reflectedAndroidString("android.os.Build\$VERSION", "RELEASE", "unknown"),
            PLATFORM_KEY to "android",
        )

    private fun reflectedAndroidString(
        className: String,
        fieldName: String,
        fallback: String,
    ): String =
        try {
            Class.forName(className).getField(fieldName).get(null) as? String ?: fallback
        } catch (_: Throwable) {
            when (fieldName) {
                "MODEL" -> System.getProperty("os.name") ?: fallback
                "RELEASE" -> System.getProperty("os.version") ?: fallback
                else -> fallback
            }
        }
}

// SDK logger (convenience wrapper)

/**
 * Simple logger for SDK components with category-based filtering.
 * Matches Swift SDK SDKLogger struct.
 */
class SDKLogger(
    val category: String = "SDK",
) {
    // Logging methods

    /**
     * Log a trace-level message (proto LOG_LEVEL_TRACE — the most verbose
     * severity, below DEBUG). No-op in release builds of the SDK, gated the
     * same way as [debug].
     */
    fun trace(
        message: String,
        metadata: Map<String, Any?>? = null,
    ) {
        if (!BuildConfig.DEBUG) return
        Logging.log(
            level = LogLevel.LOG_LEVEL_TRACE,
            category = category,
            message = message,
            metadata = metadata,
        )
    }

    /**
     * Log a debug-level message. No-op in release builds of the SDK,
     * mirroring Swift's `#if DEBUG` gate.
     */
    fun debug(
        message: String,
        metadata: Map<String, Any?>? = null,
    ) {
        if (!BuildConfig.DEBUG) return
        Logging.log(
            level = LogLevel.LOG_LEVEL_DEBUG,
            category = category,
            message = message,
            metadata = metadata,
        )
    }

    /**
     * Log an info-level message.
     */
    fun info(
        message: String,
        metadata: Map<String, Any?>? = null,
    ) {
        Logging.log(
            level = LogLevel.LOG_LEVEL_INFO,
            category = category,
            message = message,
            metadata = metadata,
        )
    }

    /**
     * Log a warning-level message.
     */
    fun warning(
        message: String,
        metadata: Map<String, Any?>? = null,
    ) {
        Logging.log(
            level = LogLevel.LOG_LEVEL_WARNING,
            category = category,
            message = message,
            metadata = metadata,
        )
    }

    /**
     * Alias for warning to match common conventions.
     */
    fun warn(
        message: String,
        metadata: Map<String, Any?>? = null,
    ) = warning(message, metadata)

    /**
     * Log an error-level message.
     */
    fun error(
        message: String,
        metadata: Map<String, Any?>? = null,
        throwable: Throwable? = null,
    ) {
        val errorMetadata =
            if (throwable != null) {
                (metadata ?: emptyMap()) +
                    mapOf(
                        "exception_type" to throwable::class.simpleName,
                        "exception_message" to throwable.message,
                    )
            } else {
                metadata
            }

        Logging.log(
            level = LogLevel.LOG_LEVEL_ERROR,
            category = category,
            message = message,
            metadata = errorMetadata,
        )
    }

    /**
     * Log a fault-level message (critical system errors). Mirrors Swift's
     * top severity (`LogLevel.fault`).
     */
    fun fault(
        message: String,
        metadata: Map<String, Any?>? = null,
        throwable: Throwable? = null,
    ) {
        val faultMetadata =
            if (throwable != null) {
                (metadata ?: emptyMap()) +
                    mapOf(
                        "exception_type" to throwable::class.simpleName,
                        "exception_message" to throwable.message,
                    )
            } else {
                metadata
            }

        Logging.log(
            level = LogLevel.LOG_LEVEL_FATAL,
            category = category,
            message = message,
            metadata = faultMetadata,
        )
    }

    // Error logging with context

    /**
     * Log an error with source location context.
     */
    fun logError(
        error: Throwable,
        additionalInfo: String? = null,
        file: String? = null,
        line: Int? = null,
        function: String? = null,
    ) {
        val errorMessage =
            buildString {
                append(error.message ?: error::class.simpleName)
                if (file != null || line != null || function != null) {
                    append(" at ")
                    file?.let { append(it) }
                    line?.let { append(":$it") }
                    function?.let { append(" in $it") }
                }
                additionalInfo?.let { append(" | Context: $it") }
            }

        val metadata =
            buildMap<String, Any?> {
                file?.let { put("source_file", it) }
                line?.let { put("source_line", it) }
                function?.let { put("source_function", it) }
                put("exception_type", error::class.simpleName)
                put("exception_message", error.message)
            }

        Logging.log(
            level = LogLevel.LOG_LEVEL_ERROR,
            category = category,
            message = errorMessage,
            metadata = metadata,
            file = file,
            line = line,
            function = function,
        )
    }

    /**
     * Log with model context (for model-related operations).
     */
    fun logModelInfo(
        message: String,
        modelId: String,
        framework: String? = null,
        metadata: Map<String, Any?>? = null,
    ) {
        Logging.log(
            level = LogLevel.LOG_LEVEL_INFO,
            category = category,
            message = message,
            metadata = metadata,
            modelId = modelId,
            framework = framework,
        )
    }

    /**
     * Log model error with context.
     */
    fun logModelError(
        message: String,
        modelId: String,
        framework: String? = null,
        errorCode: Int? = null,
        metadata: Map<String, Any?>? = null,
    ) {
        Logging.log(
            level = LogLevel.LOG_LEVEL_ERROR,
            category = category,
            message = message,
            metadata = metadata,
            modelId = modelId,
            framework = framework,
            errorCode = errorCode,
        )
    }

    // Companion object — convenience loggers

    companion object {
        /**
         * Set the global minimum log level.
         */
        fun setLevel(level: LogLevel) {
            Logging.setMinLogLevel(level)
        }

        // CONVENIENCE LOGGERS (matching Swift SDK and runanywhere-commons)

        /** Shared logger for general SDK usage */
        val shared = SDKLogger("RunAnywhere")

        /** Logger for LLM operations */
        val llm = SDKLogger("LLM")

        /** Logger for STT (Speech-to-Text) operations */
        val stt = SDKLogger("STT")

        /** Logger for TTS (Text-to-Speech) operations */
        val tts = SDKLogger("TTS")

        /** Logger for VAD (Voice Activity Detection) operations */
        val vad = SDKLogger("VAD")

        /** Logger for model management operations */
        val models = SDKLogger("Models")

        /** Logger for ONNX runtime operations */
        val onnx = SDKLogger("ONNX")

        /** Logger for LlamaCpp operations */
        val llamacpp = SDKLogger("LlamaCpp")

        /** Logger for RAG (Retrieval-Augmented Generation) operations */
        val rag = SDKLogger("RAG")

        /** Logger for VoiceAgent operations */
        val voiceAgent = SDKLogger("VoiceAgent")
    }
}
